Sources/NIO/SelectorEpoll.swift

```diff
@@ -93,6 +93,35 @@ extension SelectorEventSet {
     }
 }
 
+@usableFromInline struct EPollUserData {
+    // We're backing this by a tuple to get standard C layout
+    @usableFromInline var registrationID: SelectorRegistrationID
+    @usableFromInline var fileDescriptor: CInt
+
+
+    @inlinable init(registrationID: SelectorRegistrationID, fileDescriptor: CInt) {
+        assert(MemoryLayout<UInt64>.size == MemoryLayout<EPollUserData>.size)
+        self.registrationID = registrationID
+        self.fileDescriptor = fileDescriptor
+    }
+
+    @inlinable init(rawValue: UInt64) {
+        let registrationID = UInt32(truncatingIfNeeded: (rawValue & (UInt64(UInt32.max) << 32)) >> 32)
+        let fileDescriptor = CInt(truncatingIfNeeded: rawValue & UInt64(UInt32.max))
+
+        self = .init(registrationID: SelectorRegistrationID(rawValue: registrationID),
+                     fileDescriptor: fileDescriptor)
+    }
+}
+
+extension UInt64 {
+    @inlinable
+    init(_ epollUserData: EPollUserData) {
+        self = UInt64((UInt64(epollUserData.registrationID.rawValue) << 32) |
+                        (UInt64(UInt32(bitPattern: epollUserData.fileDescriptor))))
```

Comment:
I know we discussed what should happen if we _receive_ `-1` from the kernel, but should it be possible for us to _pass_ `-1` to it? I think I'd rather the answer was no, so I kinda think that instead of `UInt32(bitPattern:)` here we should just promote to `UInt64` and crash if this is negative.

Reply:
thanks, fixed.